File: src/src/main/resources/mmlspark/UnicodeNormalize.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *

@inherit_doc
class UnicodeNormalize(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        form (str): Unicode normalization form: NFC, NFD, NFKC, NFKD
        inputCol (str): The name of the input column
        lower (bool): Lowercase text
        outputCol (str): The name of the output column
    """

    @keyword_only
    def __init__(self, form=None, inputCol=None, lower=None, outputCol=None):
        super(UnicodeNormalize, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.UnicodeNormalize")
        self.form = Param(self, "form", "form: Unicode normalization form: NFC, NFD, NFKC, NFKD")
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.lower = Param(self, "lower", "lower: Lowercase text")
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column")
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, form=None, inputCol=None, lower=None, outputCol=None):
        """
        Set the (keyword only) parameters

        Args:

            form (str): Unicode normalization form: NFC, NFD, NFKC, NFKD
            inputCol (str): The name of the input column
            lower (bool): Lowercase text
            outputCol (str): The name of the output column
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setForm(self, value):
        """

        Args:

            form (str): Unicode normalization form: NFC, NFD, NFKC, NFKD

        """
        self._set(form=value)
        return self


    def getForm(self):
        """

        Returns:

            str: Unicode normalization form: NFC, NFD, NFKC, NFKD
        """
        return self.getOrDefault(self.form)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setLower(self, value):
        """

        Args:

            lower (bool): Lowercase text

        """
        self._set(lower=value)
        return self


    def getLower(self):
        """

        Returns:

            bool: Lowercase text
        """
        return self.getOrDefault(self.lower)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column
        """
        return self.getOrDefault(self.outputCol)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.UnicodeNormalize"

    @staticmethod
    def _from_java(java_stage):
        module_name=UnicodeNormalize.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".UnicodeNormalize"
        return from_java(java_stage, module_name)

File: core/src/main/scala/com/microsoft/ml/spark/stages/UnicodeNormalize.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.ml.spark.stages

import org.apache.spark.ml.{ComplexParamsReadable, ComplexParamsWritable, Transformer}
import org.apache.spark.ml.param.{BooleanParam, Param, ParamMap}
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.sql.{DataFrame, Dataset}
import org.apache.spark.sql.functions.udf

import java.text.Normalizer
import com.microsoft.ml.spark.codegen.Wrappable
import com.microsoft.ml.spark.core.contracts.{HasInputCol, HasOutputCol}
import com.microsoft.ml.spark.logging.BasicLogging
import org.apache.spark.sql.types.{StringType, StructField, StructType}

object UnicodeNormalize extends ComplexParamsReadable[UnicodeNormalize]

/** <code>UnicodeNormalize</code> takes a dataframe and normalizes the unicode representation.
  */
class UnicodeNormalize(val uid: String) extends Transformer
  with HasInputCol with HasOutputCol with Wrappable with ComplexParamsWritable with BasicLogging {
  logClass()

  def this() = this(Identifiable.randomUID("UnicodeNormalize"))

  val form = new Param[String](this, "form", "Unicode normalization form: NFC, NFD, NFKC, NFKD")

  /** @group getParam */
  def getForm: String = get(form).getOrElse("NFKD")

  /** @group setParam */
  def setForm(value: String): this.type = {
    // check input value
    Normalizer.Form.valueOf(getForm)

    set("form", value)
  }

  val lower = new BooleanParam(this, "lower", "Lowercase text")

  /** @group getParam */
  def getLower: Boolean = get(lower).getOrElse(true)

  /** @group setParam */
  def setLower(value: Boolean): this.type = set("lower", value)

  /** @param dataset - The input dataset, to be transformed
    * @return The DataFrame that results from column selection
    */
  override def transform(dataset: Dataset[_]): DataFrame = {
    logTransform[DataFrame]({
      val inputIndex = dataset.columns.indexOf(getInputCol)

      require(inputIndex != -1, s"Input column $getInputCol does not exist")

      val normalizeFunc = (value: String) =>
        if (value == null) null
        else Normalizer.normalize(value, Normalizer.Form.valueOf(getForm))

      val f = if (getLower)
        (value: String) => Option(value).map(s => normalizeFunc(s.toLowerCase)).orNull
      else
        normalizeFunc

      val textMapper = udf(f)

      dataset.withColumn(getOutputCol, textMapper(dataset(getInputCol)).as(getOutputCol))
    })
  }

  def transformSchema(schema: StructType): StructType = {
    schema.add(StructField(getOutputCol, StringType))
  }

  def copy(extra: ParamMap): UnicodeNormalize = defaultCopy(extra)

}
